models.py

```diff
@@ -210,85 +210,166 @@ def empty_order(self):
+        database. It also updates Items' availability.
+        :param dict items: {<Item>: <quantity:int>}
+        :param bool update_total: if True the procedure updates order's
+            total price
+        :param Address new_address: if not None the procedure updates the
+            order with the given address
+        """
+        to_create = {}
+        to_remove = {}
+        to_edit = {}
+        total_price_difference = 0
+
+        # split items in insert, delete and update sets
+        for item, quantity in items.items():
+            for orderitem in self.order_items:
+                if orderitem.item == item:
+                    difference = quantity - orderitem.quantity
+                    if quantity == 0:
+                        to_remove[item] = orderitem.quantity
+                    elif difference > item.availability:
+                        raise InsufficientAvailabilityException(
+                            item, quantity)
+                    elif quantity < 0:
+                        raise WrongQuantity
+                    else:
+                        to_edit[item] = difference
+                    total_price_difference += (item.price * difference)
+                    break
+            else:
+                if quantity <= 0:
+                    raise WrongQuantity
+                elif quantity > item.availability:
+                    raise InsufficientAvailabilityException(
+                        item, quantity)
+                else:
+                    to_create[item] = quantity
+                total_price_difference += (item.price * quantity)
+
+        with database.atomic():
+            self.edit_items_quantity(to_edit)
+            self.create_items(to_create)
+            self.delete_items(to_remove)
+            if update_total:
+                self.total_price += total_price_difference
+            if new_address:
+                self.address = new_address
+            if update_total or new_address:
+                self.save()
+        return self
 
-        :param item Item: instance of models.Item
+    def edit_items_quantity(self, items):
         """
-        for orderitem in self.order_items:
-            # Looping all the OrderItem related to this order, if one with the
-            # same item is found we update that row.
-            if orderitem.item == item:
-                orderitem.add_item(quantity)
+        Update orderitems using a query for each item, and updates
+        items' availability.
+        :param dict items: {<Item>: <difference:int>}
+        """
+        if not items:
+            return
 
-                self.total_price += (item.price * quantity)
+        with database.atomic():
+            for item, difference in items.items():
+                item.availability -= difference
+                item.save()
+                orderitem = OrderItem.get(
+                    OrderItem.item == item, OrderItem.order == self)
```

Comment:
non si poteva fare una select con un `in` invece che una get per ogni orderitem?